### PR TITLE
WIP initial auto jenkins provisioning

### PR DESCRIPTION
**Pull request by gdams**
This probably shouldn't be run until we have standardised our hostnames as it could create a load of rogue jenkins nodes that we probably wouldn't want

### CODE REVIEW

**Comment by gdams** [comment]
FYI https://ci.adoptopenjdk.net/computer/build-packet-freebsd11-x64-1/ is an example of a node that configured using this playbook:

labels: `amd64 build FreeBSD FreeBSD11`

TODO:
- Add `JDK7_BOOT_DIR`, `JDK8_BOOT_DIR`, `JDK9_BOOT_DIR`.

I feel that this can be merged without the JAVA dirs for now and they can be manually added, I will work on automating this at a later date

File: ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Jenkins_Master/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  uri:
+    url: "https://ci.adoptopenjdk.net/computer/{{ ansible_hostname }}/api/json"
+  register: jenkins
+  ignore_errors: true
+  tags: jenkins_master
+
+- name: SSH into the Jenkins and excute the jenkins-create-node.sh script
+  become: false
+  # TODO fetch a var instead of hard coding linux
+  local_action: command ssh -o StrictHostKeyChecking=no jenkins@ci.adoptopenjdk.net "/home/jenkins/.jenkins/jenkins-create-node.sh {{ inventory_hostname }} {{ ansible_host }} /home/jenkins linux,{{os}},{{arch}},{{provider}},{{ansible_architecture}},{{type}},{{ansible_distribution|lower}},{{ansible_distribution|lower}}{{ansible_distribution_major_version}}"
```
Comment (bblondin) [inline ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Jenkins_Master/tasks/main.yml]: These should be vars

```
jenkins@ci.adoptopenjdk.net --> {{ Jenkins_Username }}@{{ Jenkins_Server }}

/home/jenkins (x2) --> /home/{{ Jenkins_Username }}
```

File: ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Jenkins_Master/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+---
+- name: Check if machine exists in jenkins
+  uri:
+    url: "https://ci.adoptopenjdk.net/computer/{{ ansible_hostname }}/api/json"
```
Comment (bblondin) [inline ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Jenkins_Master/tasks/main.yml]: This should be a var

```
url: "https://ci.adoptopenjdk.net/computer/{{ ansible_hostname }}/api/json" -->
"https://{{ Jenkins_Server }}/computer/{{ ansible_hostname }}/api/json"

```

File: ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Jenkins_Master/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+---
```
Comment (bblondin) [inline ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Jenkins_Master/tasks/main.yml]:
```
##################
# Jenkins_Master #
##################
````

File: ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Nagios_Master_Config/tasks/main.yml [hunk 4]
```diff
@@ -17,7 +17,7 @@
 # Script must exist on Nagios Master
 #
 - name: SSH into the Nagios Master and excute the Nagios_Ansible_Config_tool.sh script
-  local_action: command ssh -o StrictHostKeyChecking=no root@{{ Nagios_Master_IP }} "/usr/local/nagios/Nagios_Ansible_Config_tool/Nagios_Ansible_Config_tool.sh  {{ ansible_distribution }} {{ ansible_architecture }} {{ inventory_hostname }} {{ ansible_default_ipv4.address }} "
+  local_action: command ssh -o StrictHostKeyChecking=no root@{{ Nagios_Master_IP }} "/usr/local/nagios/Nagios_Ansible_Config_tool/Nagios_Ansible_Config_tool.sh  {{ ansible_distribution }} {{ ansible_architecture }} {{ inventory_hostname }} {{ ansible_host }} "
```
Comment (bblondin) [inline ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Nagios_Master_Config/tasks/main.yml]: Nagios will require the IP address of the client machine
Which out it, Nagios will not be able to find the client machine by unresolvable hostnames 
``` abc123.adoptopenjdk.net ```

Comment (gdams) [reply ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Nagios_Master_Config/tasks/main.yml]: ansible_host is the IP address that is set in the yml inventory :)

**Comment by bblondin** [comment]
This is great work, fully automating the Jeknins end of things

**Comment by karianna** [comment]
@gdams Have the hostnames been standardised so that this PR is safe to merge?

**Comment by gdams** [comment]
I'm going to leave this open until I've worked out how to set the java paths as vars

**Comment by gdams** [comment]
I've now added java paths for Ubuntu **ONLY**.

**Comment by gdams** [comment]
I'm going to bring @smlambert into this conversation because I want to make sure that the testing team is happy with the labels that are generated by this PR. I know that there was an issue that Shelley raised about standardising them but I'm not sure if this schema will work.

An example test machine would have the following labels:
`linux ubuntu1604 x64 cloudcone x86_64 test ubuntu ubuntu16`

**Comment by smlambert** [comment]
Yes, we have started using some of the labels discussed in #93, particularly our Jenkins scripts look for sw.os.xxx, and hw.arch.xxx.  Thanks @gdams for noticing that I had added those labels manually to the previous 2 x86-64_linux machines we have for test.

The reason I need/want the standard naming is so that downstream CI servers and these AdoptOpenJDK servers (ci.adopt and ci-jck.adopt) align in a labelling strategy, and we can directly reuse the Jenkins collateral from AdoptOpenJDK.  The more alignment we have, the more people we can bring to the AdoptOpenJDK project, as the work here is then directly reusable.

**Comment by gdams** [comment]
thanks for that @smlambert. Is the above schema that I mentioned going to be accpetable for you or am I going to need to change it? Basically I am going to either have to modify all of the test jobs or all of the build jobs

**Comment by smlambert** [comment]
I think we should get final agreement on #93 and have all machines following the agreed upon schema, I can help update what ever build jobs would need it.

I also am suggesting we rethink how we are using the machines, and whether we should consider having them configured for dual purpose, see https://github.com/AdoptOpenJDK/openjdk-infrastructure/issues/199#issuecomment-370237489

If the requirement is as @karianna says "each type of build does need to run on an identical host for consistency and auditability", then build machines can be statically allocated, but to avoid them sitting idle, all but one of each platform of build machines could be labelled as ci.role.test and ci.role.build, so that when they are idle, they go into the pool of machines available to test jobs (keeping 1 as exclusively ci.role.build, is the cheap way of ensuring builds never wait for tests to finish, though there are also Jenkins plugins that will prioritize certain jobs over others, that could be combined for a hybrid usage story).

**Comment by karianna** [comment]
I think we can loosen the requirement of identical host as long as the libraries (native or otherwise) that we build with are identical.  If we deviate from that, then we need to make that clear to the end user (again that *might* be acceptable.

**Comment by jdekonin** [comment]
Can we please have this as a role that is enabled via option and not on as a default?  There are users that would like to consume these scripts and not have to add to a jenkins server.

**Comment by karianna** [comment]
@gdams there's conflicts.

**Comment by gdams** [comment]
I think that this was being closed in favour of https://github.com/AdoptOpenJDK/openjdk-jenkins-helper but I'll let someone confirm that before closing

**Comment by smlambert** [comment]
I think it is worth having an infra call about this, so we can discuss requirements & options.  There are varying approaches and opinions, and It will be good to converge.

**Comment by karianna** [comment]
@smlambert Yep - I'm meeting George and John online tomorrow at 1900 UK time if maybe after that  suits?

**Comment by sxa555** [comment]
@karianna What was the outcome of that meeting in September 2018? ;-)

**Comment by karianna** [comment]
> @karianna What was the outcome of that meeting in September 2018? ;-)

Hmmm, yes well.  Let's have a new call between @gdams and @smlambert on this one.

@gdams - If it's trivial you might as well solve the merge conflict and address @jdekonin's comment as well.

**Pull request closed without merging**